### PR TITLE
support multi currency rewards

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

**Review by wangjj9219 — APPROVED** [review]

File: rewards/src/lib.rs [hunk 1]
```diff
@@ -1,25 +1,31 @@
...
 	fmt::Debug,
+	vec::Vec,
 };
 
-use orml_traits::RewardHandler;
-
 /// The Reward Pool Info.
 #[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, Default, MaxEncodedLen)]
-pub struct PoolInfo<Share: HasCompact, Balance: HasCompact> {
+pub struct PoolInfoV0<Share: HasCompact, Balance: HasCompact> {
```
Comment (xlc) [inline rewards/src/lib.rs:28]: I think we can move this to migration

File: rewards/src/lib.rs [hunk 2]
```diff
@@ -1,25 +1,31 @@
...
-	FixedPointNumber, FixedPointOperand, FixedU128, RuntimeDebug,
+	traits::{AtLeast32BitUnsigned, MaybeSerializeDeserialize, Member, Saturating, UniqueSaturatedInto, Zero},
+	FixedPointOperand, RuntimeDebug, SaturatedConversion,
 };
 use sp_std::{
+	borrow::ToOwned,
 	cmp::{Eq, PartialEq},
+	collections::btree_map::BTreeMap,
 	fmt::Debug,
+	vec::Vec,
```
Comment (xlc) [inline rewards/src/lib.rs:23]: should have `prelude::*` which includes vec and few others

File: rewards/src/lib.rs (hunk 1 above)
Comment (ermalkaleci) [reply rewards/src/lib.rs:28]: ok

File: rewards/src/lib.rs [hunk 3]
```diff
@@ -124,11 +130,16 @@ pub mod module {
...
+		reward_currency: T::CurrencyId,
+		reward_increment: T::Balance,
+	) -> DispatchResult {
 		if reward_increment.is_zero() {
-			return;
+			return Ok(());
 		}
-		Pools::<T>::mutate_exists(pool, |maybe_pool_info| {
+		Pools::<T>::mutate_exists(pool, |maybe_pool_info| -> DispatchResult {
+			ensure!(maybe_pool_info.is_some(), Error::<T>::PoolDoesNotExist);
```
Comment (xlc) [inline rewards/src/lib.rs:142]: `let Some(pool_info) = maybe_pool_info.ok_or(Error::<T>::PoolDoesNotExist)?`

Comment (ermalkaleci) [reply rewards/src/lib.rs:142]: > let pool_info = maybe_pool_info.ok_or(Error::<T>::PoolDoesNotExist)?;

missed that 👌

File: rewards/src/migrations.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			});
+		}
+		reads_writes += 1;
+	}
+
+	// Return the weight consumed by the migration.
+	T::DbWeight::get().reads_writes(reads_writes, reads_writes)
+}
+
+// #[test]
```
Comment (xlc) [inline rewards/src/migrations.rs:72]: what happen to this?

**Review by xlc — COMMENTED** [review]
Should have some pre_migrate & post_migrate to allow us to verify state with try-runtime

https://github.com/paritytech/substrate/blob/master/frame/membership/src/migrations/v4.rs#L83

File: rewards/src/migrations.rs (hunk 4 above)
Comment (wangjj9219) [reply rewards/src/migrations.rs]: triggered by migrate process of module-incentives

**Review by xlc — APPROVED** [review]

**Pull request merged**